No-credential connections and 0-RTT fixes

jagerman opened the pull request:
This commit adds the ability for a client to connect to a server without providing a key, and allows listening servers to choose whether client keys are required, requested, or omitted.  Required means a client must provide a key, and it must pass the key verification function; requested means the server will request a key and the client may provide one: the key verification function is called with an empty key if the client does not provide one (so that it can still use the ALPN to reject the connection, for example, if certain ALPNs require keys and others don't).  Omitted means the server won't request one at all.

For omitted client pubkeys (either not requested, or requested and not provided but still allowed), the connection's `remote_key()` will return an empty span.

Client connections can now use GNUTLSCreds::make_unauthenticated to make a connection without a pubkey.  They can also omit the TLSCreds from the connect() call entirely, but that won't allow doing things like enable 0rtt.  This connection with such a missing pubkey will fail if the server requires pubkeys, but otherwise is allowed.

This deliberately renames `set_key_verify_callback` to `require_client_keys` (thus breaking code) as calling code needs to decide whether it wants requires or requests behaviour.

This also fixes closely related issues with 0-RTT: under 0-RTT resumption, key verification callbacks were never called, and so the remote_key was never set on the connection object.  This is actually intentional: a valid 0-RTT ticket already contains the connection data, and so the initial connection that produced the ticket already passed verification.  This commit contains fixes to this so that they get properly reflected in the Connection object, and documents this for the key verification functions.

Other changes:

- remove pointless `gtls_key` struct and make it back into a simple typedef, like it used to be.  The indirection added nothing.

- replace Connection `is_validated()` with `is_handshaked()` and `is_handshake_confirmed()`.  Validated isn't used anywhere outside of libquic's test suite, and doesn't really work with 0-RTT connections that don't re-validate.

- remove `set_draining()` and `set_closing()` -- they were pointless private methods that simply set the variable of the same name to false.  But anything with private access can already just set those to false itself.

- Fixed wrong comment above cert_verify_callback_gnutls about what a failure return value looks like.  The function was already returning 1 rather than a negative (which is actually okay), but this intent is now clearer in the comment and the code.

- Removed unused `usage_flags`.  It was already tagged `[[maybe_unused]]` but was actually unused and didn't seem to have any purpose.

jagerman committed:
No-credential connections and 0-RTT fixes

This commit adds the ability for a client to connect to a server without
providing a key, and allows listening servers to choose whether client
keys are required, requested, or omitted.  Required means a client must
provide a key, and it must pass the key verification function; requested
means the server will request a key and the client may provide one: the
key verification function is called with an empty key if the client does
not provide one (so that it can still use the ALPN to reject the
connection, for example, if certain ALPNs require keys and others
don't).  Omitted means the server won't request one at all.

For omitted client pubkeys (either not requested, or requested and not
provided but still allowed), the connection's `remote_key()` will return
an empty span.

Client connections can now use GNUTLSCreds::make_unauthenticated to make
a connection without a pubkey.  They can also omit the TLSCreds from the
connect() call entirely, but that won't allow doing things like enable
0rtt.  This connection with such a missing pubkey will fail if the
server requires pubkeys, but otherwise is allowed.

This deliberately renames `set_key_verify_callback` to
`require_client_keys` (thus breaking code) as calling code needs to
decide whether it wants requires or requests behaviour.

This also fixes closely related issues with 0-RTT: under 0-RTT
resumption, key verification callbacks were never called, and so the
remote_key was never set on the connection object.  This is actually
intentional: a valid 0-RTT ticket already contains the connection data,
and so the initial connection that produced the ticket already passed
verification.  This commit contains fixes to this so that they get
properly reflected in the Connection object, and documents this for the
key verification functions.

Other changes:

- remove pointless `gtls_key` struct and make it back into a simple
  typedef, like it used to be.  The indirection added nothing.

- replace Connection `is_validated()` with `is_handshaked()` and
  `is_handshake_confirmed()`.  Validated isn't used anywhere outside of
  libquic's test suite, and doesn't really work with 0-RTT connections
  that don't re-validate.

- remove `set_draining()` and `set_closing()` -- they were pointless
  private methods that simply set the variable of the same name to
  false.  But anything with private access can already just set those to
  false itself.

- Fixed wrong comment above cert_verify_callback_gnutls about what a
  failure return value looks like.  The function was already returning 1
  rather than a negative (which is actually okay), but this intent is
  now clearer in the comment and the code.

- Removed unused `usage_flags`.  It was already tagged
  `[[maybe_unused]]` but was actually unused and didn't seem to have any
  purpose.

tewinget commented on include/oxen/quic/gnutls_crypto.hpp:
```diff
@@ -367,10 +351,17 @@ namespace oxen::quic
+        // Construct a credentials object from a Ed25519 combined seed/pubkey value.
         static std::shared_ptr<GNUTLSCreds> make_from_ed_seckey(std::string_view sk);
+        // Constructs a no-credentials object.  This object may *only* be used for client
+        // connections, and will fail if the server is configured to require a client certificate.
```
s/and will fail/and connections using it will fail/

jagerman committed:
Don't try to get ALPN before conn is handshaked

Application callbacks should be able to call `selected_alpn()` without
having to worry about it throwing for not-yet-established connections.

jagerman committed:
improve comment wording

jagerman committed:
Don't fire close callback on failed incoming connections

If an incoming connection closed before handshaking is completed, we are
currently calling the connection close callback, but that doesn't make
sense because we never fired the connection established callback, and so
the application has never been told about this connection.

This changes is so that upon such failures we don't fire the connection
closed callback and just let it die silently.

jagerman committed:
Check for no certs before raw pubkey mode

If client doesn't send a cert then the raw pk mode never gets
negotiated, producing warnings here.

jagerman committed:
Fix queue_incoming_stream when called after a stream is created

Currently if an incoming stream is created without a queued stream and
*then* queue_incoming_stream is called, the queued stream would get
assigned id 0, but that id is already used by the original stream, and
so the queued stream could never get used.

jagerman committed:
Fire conn established earlier for 0-RTT connection

Applications (e.g. lokinet) typically do things in the conn_established
callback, but for a 0-RTT server that receives early stream data or
datagrams, the stream gets created *before* the conn_established
callback, which is unintuitive and counterproductive.

This corrects it so that if we do get early data, we fire the conn
established callback immediately before opening to stream so that
whatever initialization is needed (such as queuing an incoming stream)
can be done safely.

tewinget commented on src/connection.cpp:
```diff
@@ -1354,6 +1354,8 @@ namespace oxen::quic
             assert(ins);
             return 0;
         }
+        else if (id == next_incoming_stream_id)
```
This will work if we only ever allow bidirectional streams.  I wonder if the better answer is to have two queues for incoming streams (uni and bidi) and not assign an id to them at all, but rather if there is a pending incoming stream on the queue when a stream is opened, we pop it, use it, and assign its id at that time.  I don't *think* anything is relying on that queued stream having an id set before it's opened.

jagerman replied:
Just to leave a comment here about why this is tracking IDs:

The reason for tracking the stream is to deal with ngtcp2 not calling stream opened callbacks for implicitly opened streams.

I.e. in QUIC, if you see new data on stream 8, then the QUIC protocol implies opening of stream 0 and 4.  (+1/2/3 to all of these for unidirectional and/or server-initiated), and ngtcp2 fires 8 but *declines* to fire 0 and 4 stream opened callbacks (although it is marked as subject to change in the future).  If you're trying to match up streams between client and server (i.e. server and client treat stream 0 as a BTReqStream, stream 4 as some custom stream, and stream 8 as a plain stream) then things would break as a result.

Thus tracking the ID, so that if the client opens three streams (which will always be 0-4-8) and the server queues three incoming streams (0-4-8), everything matches up without requiring ngtcp2 to call them in order.

As to unidirectional streams: yeah, this is indeed broken, but currently I don't think we handle unidirection streams at all, and we would need to either add more counters, or else just deprecate this in favour of using stream construction callbacks instead.

tewinget commented on src/connection.cpp:
```diff
@@ -495,6 +489,19 @@ namespace oxen::quic
 
+    void Connection::check_established()
+    {
+        if (establish_hook_called)
```
both call sites are also doing this check, so it's slightly redundant (but probably good to keep anyway?)

tewinget replied:
nevermind I'm blind.

jagerman replied:
There are a couple others that don't check it, and the ones that do are basically doing it to add a log statement for the "weird" case here (and the tiny gain of not making another function call).
